src/finn/transformation/fpgadataflow/ip_cache.py

```diff
@@ -0,0 +1,642 @@
+    """
+    global CACHE_IP_DEFINITIONS
+
+    def wrapper(op_cls: type) -> type:
+        assert issubclass(
+            op_cls, HWCustomOp
+        ), f"Can only cache HWCustomOp instances, but {op_cls.__name__} is not a HWCustomOP!"
+        if op_cls not in CACHE_IP_DEFINITIONS.keys():
+            CACHE_IP_DEFINITIONS[op_cls] = {}
+        else:
+            # Already marked
+            return op_cls
+        if attributes is not None:
+            CACHE_IP_DEFINITIONS[op_cls]["use"] = attributes
+        else:
+            # List of fields that don't define the IP core itself,
+            # and can thus be ignored when hashing
+            ignore_fields = [
+                "code_gen_dir_ipgen",
+                "gen_top_module",
+                "ip_vlnv",
+                "ipgen_path",
+                "ip_path",
+                "cycles_rtlsim",
+                "cycles_estimate",
+                "res_estimate",
+                "res_synth",
+                "rtlsim_so",
+                "executable_path",
+                "res_hls",
+                "code_gen_dir_cppsim",
+            ]
+            CACHE_IP_DEFINITIONS[op_cls]["ignore"] = ignore_fields
+        return op_cls
+
+    return wrapper
+
+
+class IPCache:
+    """Manage IP caching.
+
+    Public methods that are relevant for the caches usage:
+    - `model = cache.apply(model)`: Fetch cached IPs and apply them to the model,
+        returning the new model
+    - `cache.update(model)`: Update the cache by adding synthesized IPs that are not
+        yet cached into the cache.
+    - `cache.get_key(op, model)`: Get the key (string) of the given custom op
+    - `cache.get_hash_hex(key)`: Get the hex representation of the hash of the given key.
+    - `cache.get_num_cached_ips(model)`: Get the number of cached IPs in the given model.
+    """
+
+    allowed_hashfuncs: Final[list[str]] = ["sha256", "sha512", "blake2s", "blake2b"]
+
+    def __init__(
+        self,
+        cache_dir: Path,
+        hashfunc: str,
+        hls_clk_period: float,
+        cache_hls_clk: bool,
+        fpgapart: str,
+        cache_fpgapart: bool,
+    ) -> None:
+        """Construct a new IPCache object.
+
+        Args:
+            cache_dir: The path of the cache directory.
+            hashfunc: The name of the hash function to be used.
+            hls_clk_period: HLS clock period in ns.
+            cache_hls_clk: Use the HLS clock as part of the key.
+            fpgapart: FPGA-part used for HLSSynth and PrepareIP.
+            cache_fpgapart: Use the fpgapart as part of the key.
+        """
+        self.cache_dir = cache_dir
+        self.cache_hls_clk = cache_hls_clk
+        self.cache_fpgapart = cache_fpgapart
+
+        # Used to check validity of cache directory names
+        if sys.platform != "win32":
+            self.max_hash_len = os.pathconf("/", "PC_NAME_MAX")
+            self.max_path_len = os.pathconf("/", "PC_PATH_MAX")
+        else:
```

Comment:
We don't need to support Windows at all.

Reply:
I wouldn't bother at all with this. We assume a very specific Ubuntu setup already with heavy tool chains and a long list of apt packages installed already. We can assume every user has read the basic quickstart documentation.